kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/format/KafkaEntryFormatter.java

```diff
@@ -78,23 +77,24 @@ public DecodeResult decode(List<Entry> entries, byte magic) {
 
                     // batch magic greater than the magic corresponding to the version requested by the client
                     // need down converted
-                    if (batchMagic > magic) {
+                    if (batchMagic > magic || batchMagic != RecordBatch.MAGIC_VALUE_V2) {
                         MemoryRecords memoryRecords = MemoryRecords.readableRecords(ByteBufUtils.getNioBuffer(byteBuf));
                         //down converted, batch magic will be set to client magic
                         ConvertedRecords<MemoryRecords> convertedRecords =
-                                memoryRecords.downConvert(magic, startOffset, time);
+                                KopRecordsUtil.convertAndAssignOffsets(memoryRecords.batches(), magic, startOffset);
```

Comment:
We need to verify the internal message offset whether it is down-conversion or when directly returning v0 and v1 messages. Since we did not ensure the internal offset in production, we need to support it here.